Link on install

roigcarlo opened the pull request:
Cleanup from #5145 and #5170 

This is a draft, I am still testing everything, I just want some CI to trigger as well

List of changes:
- [x] New install mechanism is implemented for all applications
- [x] CMake files have been harmonized between all applications
- [x] Tests are now mandatory. Applications without tests now have the automatically generated ones with dummy tests.
- [x] **TrilinosApplication**: Some scripts are installed out of the new mechanism. Still testing those.
- [x] **CoSimulationApplication**: I did not understand the CMake.
- [x] Improve readme and put some doc into the configuration files.

Commits are done per application, so if any application needs to remain with the old mechanism ( you won't be able to install with the rest ) it still can be reverted.

RiccardoRossi committed:
prototype installation with links

RiccardoRossi committed:
adding functions

roigcarlo committed:
Install always

roigcarlo committed:
Cleanup, default install paths

roigcarlo committed:
INSTALL_LINK -> INSTALL_PYTHON_USING_LINKS

roigcarlo committed:
Starting to remove KRATOS_INSTALL_PREFIX

roigcarlo committed:
KRATOS_INSTALL_PREFIX -> CMAKE_INSTALL_PREFIX

roigcarlo committed:
Remove comments

roigcarlo committed:
Fixing error installyng python __init__.py files

roigcarlo committed:
Cleanup

roigcarlo committed:
Adapting CSharpWrapperApplication CMake to new install mechanism

roigcarlo committed:
Adapting CableNetApplication CMake to new install mechanism

roigcarlo committed:
Adapting CompressiblePotentialFlowApplication CMake to new install mechanism

roigcarlo committed:
Adapting ConstitutiveModelsApplication CMake to new install mechanism

roigcarlo committed:
Adapting ContactMechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting ContactStructuralMechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting ConvectionDiffusionApplication CMake to new install mechanism

roigcarlo committed:
Adapting DEMApplication CMake to new install mechanism

roigcarlo committed:
Adapting DamApplication CMake to new install mechanism

roigcarlo committed:
Adapting DelaunayMeshingApplication CMake to new install mechanism

roigcarlo committed:
Adapting DemStructuresCouplingApplication CMake to new install mechanism

roigcarlo committed:
Adapting EigenSolversApplication CMake to new install mechanism

roigcarlo committed:
Adapting EmpireApplication CMake to new install mechanism

roigcarlo committed:
Adapting ExaquteSandboxApplication CMake to new install mechanism

roigcarlo committed:
Adapting ExternalSolversApplication CMake to new install mechanism

roigcarlo committed:
Adapting FSIApplication CMake to new install mechanism

roigcarlo committed:
Adapting FemToDemApplication CMake to new install mechanism

roigcarlo committed:
Adapting FluidTransportApplication CMake to new install mechanism

roigcarlo committed:
Adapting FreeSurfaceApplication CMake to new install mechanism

roigcarlo committed:
Adapting HDF5Application CMake to new install mechanism

roigcarlo committed:
Adapting IgaApplication CMake to new install mechanism

roigcarlo committed:
Adapting MappingApplication CMake to new install mechanism

roigcarlo committed:
Adapting MeshMovingApplication CMake to new install mechanism

roigcarlo committed:
Adapting MeshingApplication CMake to new install mechanism

roigcarlo committed:
Adapting MetisApplication CMake to new install mechanism

roigcarlo committed:
Adapting MultilevelMonteCarloApplication CMake to new install mechanism

roigcarlo committed:
Adapting ParticleMechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting PfemApplication CMake to new install mechanism

roigcarlo committed:
Adapting PfemFluidApplication CMake to new install mechanism

roigcarlo committed:
Adapting PfemSolidApplication CMake to new install mechanism

roigcarlo committed:
Adapting PoromechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting ShallowWaterApplication CMake to new install mechanism

roigcarlo committed:
Adapting ShapeOptimizationApplication CMake to new install mechanism

roigcarlo committed:
Adapting SolidMechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting StructuralMechanicsApplication CMake to new install mechanism

roigcarlo committed:
Adapting SwimmingDEMApplication CMake to new install mechanism

roigcarlo committed:
Adapting TrilinosApplication CMake to new install mechanism

roigcarlo committed:
Adapting ULFApplication CMake to new install mechanism

roigcarlo committed:
Adapting UmatApplication CMake to new install mechanism

roigcarlo committed:
Adapting incompressible_fluid_application CMake to new install mechanism

roigcarlo committed:
Adapting pfem_2_application CMake to new install mechanism

roigcarlo committed:
Fixing merge problem in FluidDynamics CMake

roigcarlo committed:
adding generic test files for external and metis

roigcarlo committed:
Updating templates

roigcarlo committed:
Updating newly added tests of ExternalSolversApplication to follow template

roigcarlo committed:
Updating newly added tests of MetisApplication to follow template

roigcarlo committed:
Added templated tests for DamApplication

roigcarlo committed:
Added templated tests for FemToDemApplication

roigcarlo committed:
Added templated tests for PfemSolidMechanicsApplication

roigcarlo committed:
Added templated tests for ULFApplication

roigcarlo committed:
Added templated tests for pfem_2_application

roigcarlo committed:
Added templated tests for incompressible_fluid__application

roigcarlo committed:
Updating CI's

roigcarlo commented:
It's starting to work so you can start reviewing while I fix the problems :+1:

philbucher commented:
CoSimApp:
This app also has a py-only version, but I can adapt it afterwards.
The regular cmake is still there, just guarded by an if: https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CoSimulationApplication/CMakeLists.txt#L23-L70

roigcarlo committed:
Fixing problem with the nightly tests

roigcarlo committed:
Correcting app generator template for nightly tests

roigcarlo committed:
Added templated tests for DemStructuresCouplingApplication

roigcarlo committed:
Fix typo in assertEqual

roigcarlo committed:
Typo

roigcarlo committed:
Disabling GUI tests depending on problemtype scripts

roigcarlo committed:
Installing analysis submodule

roigcarlo committed:
Merge branch 'master' into cleanup/link_on_install

roigcarlo commented:
Wii, it passes, but I still getting an error in the StructuralMechanics MPI. Is it normal? I understand that the main script is not catching the exit code because is a mpi process, but I am not sure about the error

roigcarlo commented:
> Why not recursive?

Can it be done?

philbucher commented:
> > Why not recursive?
> 
> Can it be done?

Idk, you are doing this :smile: 

To me this would be a blocker

philbucher commented:
> Wii, it passes, but I still getting an error in the StructuralMechanics MPI. Is it normal? I understand that the main script is not catching the exit code because is a mpi process, but I am not sure about the error

you can ignore it, I should update this test, sorry for the confusion :/

roigcarlo committed:
Implementing recursive install for submodules

roigcarlo commented:
Done, let's see if it works. Crossing fingers.

I am still investigating if the LINK option creates a softlink of the folder....

philbucher reviewed (commented):
does it also work for subsubmodules?
I mean really recursive?
Asking bcs I don't see the fct call for the recursive

roigcarlo committed:
Implementing recursive install for submodules with LINK option enabled

roigcarlo commented:
It should. I will retest. And now it also should work with the LINK option on.

The recursive option is in 
`file(GLOB_RECURSE SRC_FILES RELATIVE ${origin_dir} ${origin_dir}/*.py )`

It previously was:
`file(GLOB SRC_FILES ${origin_dir}/*.py )`

Basically what I do is, first iterate recursively over all the files in the folder, I get its name relative to the fetch dir, so I only get either the name of the file or the subfolder skeleton + the name. 
Then its only mater of copying the files in the right order, which is provided by CMake (or it should at least, that the part I want to double check)

For the links pretty much the same but with the caveat that `cmake -e` is not able even to create the first level of directory for subfolders, so I split the macro in order to be able to create the dir before the actual sublink is created.

roigcarlo commented:
Yep. It iterates recursively:

```bash
➜  Linux git:(cleanup/link_on_install) ✗ ls FullDebug/KratosMultiphysics/DEMApplication/analytic_tools/second_subdir
__init__.py  calculate_frustation_level.py
```

roigcarlo committed:
Now I remember why it was there.

roigcarlo commented:
@philbucher I answer you here because I don't get notifications on commit comments.

Yep we can print an error on win :+1: but I don't agree to having the Link solution as the default one for several reasons: 

- I see that option just as a convenience to ease the life of the developers if they want to touch the python scripts directly. 
- Keeping all builds completely isolated by default minimizes the chances of changes being made while you are developing something in one build affect the other. Simple example, you have debug and release config and you update an script in the source, that would affect both installations, if you have to manually copy it via `make install` you are sure about what you are modifying.

Probably I am being too enthusiastic about this, but it is something I would really really want to change. 

Anyway, I think the best will be discussing this on the meeting

RiccardoRossi commented:
ok to me for not being link by default as long as it is simple to activate

what is that needs to be copied and not linked??

On Wed, Oct 23, 2019, 11:13 AM Carlos Roig <notifications@github.com> wrote:

> @philbucher <https://github.com/philbucher> I answer you here because I
> don't get notifications on commit comments.
>
> Yep we can print an error on win 👍 but I don't agree to having the Link
> solution as the default one for several reasons:
>
>    - I see that option just as a convenience to ease the life of the
>    developers if they want to touch the python scripts directly.
>    - I am not even sure about the last as the FooApplication.py is being
>    copied as the __init__.py of the module anyway (that is not being done
>    as a link) so you end up copying some things because are necessary but
>    others not, and we don't really explain why anywhere. If its hard for us to
>    tell, imagine for someone who is installing Kratos for the first time...
>    - Keeping all builds completely isolated by default minimizes the
>    chances of changes being made while you are developing something in one
>    build affect the other. Simple example, you have debug and release config
>    and you update an script in the source, that would affect both
>    installations, if you have to manually copy it via make install you
>    are sure about what you are modifying.
>
> Probably I am being too enthusiastic about this, but it is something I
> would really really want to change.
>
> Anyway, I think the best will be discussing this on the meeting
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5825?email_source=notifications&email_token=AB5PWEKYWCYCGIMPNIM6DGTQQAIS3A5CNFSM4JDNJMG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECAV4YA#issuecomment-545349216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEIZR72XNNXDUWUTV43QQAIS3ANCNFSM4JDNJMGQ>
> .
>

roigcarlo commented:
@RiccardoRossi Nothing, I was wrong. Its also installed as a link

philbucher commented:
ok probably better to discuss in the meeting

can we at least make travis to use the links?

RiccardoRossi commented:
+1000 for using links on Travis (if it works with links the contrary is
guaranteed...)

On Wed, Oct 23, 2019, 1:29 PM Philipp Bucher <notifications@github.com>
wrote:

> ok probably better to discuss in the meeting
>
> can we at least make travis to use the links?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5825?email_source=notifications&email_token=AB5PWEOKFUG4IOJPYSWZQRLQQAYQDA5CNFSM4JDNJMG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECBCB5I#issuecomment-545399029>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEOUPTUQSVG6L5JYXBDQQAYQDANCNFSM4JDNJMGQ>
> .
>

roigcarlo commented:
Yep no problem on using links there :+1:

roigcarlo committed:
Using link install on travis

philbucher commented:
3 comments:
- Can you please update the CoSimApp?
- Why do you update the deprecated apps like the incompressible_fluid?
- I think we have to change the import mechanism. Otherwise if we still do the `__path__` manipulation, how can you know which script is being used?

pooyan-dadvand committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into cleanup/link_on_install

philbucher commented:
@roigcarlo this is what I mean, did you try with removing it?
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_interface/__init__.py#L72

roigcarlo commented:
Give me a sec, we are merging some changes from @pooyan-dadvand into this. I'll try afterwards

philbucher commented:
looking at it, I think now only [importing the app into the Kernel](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_interface/__init__.py#L75) should be left to do, the rest is done from CMake

pooyan-dadvand committed:
Fixing Eigen application

pooyan-dadvand committed:
Adding blas

pooyan-dadvand committed:
Changing to KRATOS_SOURCE_DIR

pooyan-dadvand committed:
Moving runkratos

pooyan-dadvand committed:
Updating gitignore file

pooyan-dadvand committed:
Updating CMakeLists

pooyan-dadvand committed:
Adding dependency handler

pooyan-dadvand committed:
Use include paths

roigcarlo committed:
Updating scripts and adding default one to the root dir

roigcarlo committed:
Remove old scripts

roigcarlo committed:
Merge branch 'master' into cleanup/link_on_install

roigcarlo committed:
Flags for travis

roigcarlo committed:
Merge branch 'cleanup/link_on_install' of https://github.com/KratosMultiphysics/Kratos into cleanup/link_on_install

roigcarlo committed:
Improving performance and avoid duplicates

roigcarlo committed:
Fix test path

roigcarlo commented:
@philbucher Well now it runs with the tests finally, what was the problem with the importing of the kernel again? Shall I remove that?

philbucher committed:
removing adding py-scripts to __path__

philbucher commented:
> @philbucher Well now it runs with the tests finally, what was the problem with the importing of the kernel again? Shall I remove that?

=> #5918

philbucher commented:
I have not tested it but I am quite convinced it needs to be removed

roigcarlo committed:
Merge pull request #5918 from KratosMultiphysics/core/link-on-install-impr

removing adding py-scripts to __path__

roigcarlo commented:
It does work :+1: , the only problem I see is that travis seems to copy the python links before compiling, which is odd as there are marked as install targets and not file copies :S

roigcarlo committed:
Updating scripts

roigcarlo committed:
Use gcc 7 on bionic

roigcarlo committed:
Trying to install matching fortran

roigcarlo committed:
Missing 7

roigcarlo committed:
remove verbose message

roigcarlo committed:
Appveyor build

philbucher commented on scripts/build/travis/configure_travis_trusty.sh:
```diff
@@ -1,186 +1,52 @@
+export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}${KRATOS_APP_DIR}/DEMApplication;"
+export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}${KRATOS_APP_DIR}/CSharpWrapperApplication;"
+export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}${KRATOS_APP_DIR}/MetisApplication;"
+export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}${KRATOS_APP_DIR}/TrilinosApplication;"
```
is this how the configure script will look in the future?
Looks quite complicated compared to the current situation

roigcarlo replied:
Well you can have a list of applications, like
```export KRATOS_APPLICATIONS="${KRATOS_APP_DIR}/DEMApplication;${KRATOS_APP_DIR}/CSharpWrapperApplication;${KRATOS_APP_DIR}/MetisApplication;..."``` if you prefer.

But I understand you you mean. Another solution is to have a text file with the list of applications that you want to use @pooyan-dadvand what you think about that?

roigcarlo replied:
I came up with a couple more alternatives. We can also do:

```bash
function add_app {
    export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}$1;"
}

add_app ${KRATOS_APP_DIR}/ExternalSolversApplication
add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication
# The ${KRATOS_APP_DIR} may be also hidden in the function as an optional parameter.
```

or 

```bash
APPLICATIONS=(
    "${KRATOS_APP_DIR}/ExternalSolversApplication"
    "${KRATOS_APP_DIR}/StructuralMechanicsApplication"
    "${KRATOS_APP_DIR}/FluidDynamicsApplication"
)

export KRATOS_APPLICATIONS=`printf %s";" "${APPLICATIONS[@]}"`
```

Both will work (I have to think the counterparts in win). Now that I code them I think the first one is quite clean. What do you think?

philbucher replied:
Oh ok I was not aware that this PR also changes completely how we compile Kratos :sweat_smile:

roigcarlo replied:
Yep, we are taking the chance to catch up with what altair already has as well so minimize conflicts

philbucher replied:
could you post an example of how the complete configure file (or whatever it will be) looks like then?

roigcarlo replied:
```bash
#!/bin/bash
# You can use your interpreter of choice (bash, sh, zsh, ...)

# For any question please contact with us in:
#   - https://github.com/KratosMultiphysics/Kratos

function add_app {
    export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}$1;"
}

# Set compiler
export CC=gcc
export CXX=g++

# Set variables
export KRATOS_SOURCE="${KRATOS_SOURCE:-${PWD}}"
export KRATOS_BUILD="${KRATOS_SOURCE}/build"
export KRATOS_APP_DIR="${KRATOS_SOURCE}/applications"

# Set applications to compile
export KRATOS_BUILD_TYPE=${KRATOS_BUILD_TYPE:="Release"}

add_app ${KRATOS_APP_DIR}/ExternalSolversApplication
add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication

# Clean
clear
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/cmake_install.cmake"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeCache.txt"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeFiles"

# Configure
cmake -H"${KRATOS_SOURCE}" -B"${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" \
-DINSTALL_PYTHON_USING_LINKS=ON \
-DUSE_COTIRE=ON

# Buid
cmake --build "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" --target all_unity -- -j4
cmake --build "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" --target install/fast -- -j4 
```

roigcarlo committed:
Typos

roigcarlo committed:
Corrected path

roigcarlo committed:
Copypaste errors

roigcarlo committed:
Prevent duplications

roigcarlo committed:
Adding dependencies

roigcarlo committed:
Adding warning

roigcarlo committed:
Adding sample configure for win

roigcarlo committed:
Correcting bats

roigcarlo committed:
Adding debug

roigcarlo committed:
Adding full echo

roigcarlo committed:
Adding full echo

roigcarlo committed:
Moving mesh apps?

roigcarlo committed:
Restore file

roigcarlo committed:
Avoid conflicts with the other configure file

roigcarlo committed:
Trying to understand what is happening...

roigcarlo commented:
Guys does someone have an idea why appeyor is deciding not to execute certain instructions of the script?

```cmd
rem For any question please contact with us in:
rem  - https://github.com/KratosMultiphysics/Kratos

rem Set compiler
@echo off
set CC=cl.exe
set CXX=cl.exe

rem Set variables
set KRATOS_BUILD_TYPE=Release
set KRATOS_SOURCE=.
set KRATOS_BUILD=.\build
set KRATOS_APP_DIR=applications

CALL :add_app %KRATOS_APP_DIR%\FluidDynamicsApplication;
CALL :add_app %KRATOS_APP_DIR%\StructuralMechanicsApplication;
CALL :add_app %KRATOS_APP_DIR%\ContactStructuralMechanicsApplication;
CALL :add_app %KRATOS_APP_DIR%\MeshingApplication;
CALL :add_app %KRATOS_APP_DIR%\MeshMovingApplication;
VVVVVVVVVVVVVVVVVVV
REM It stops here
^^^^^^^^^^^^^^^^^^^^^^^^
CALL :add_app %KRATOS_APP_DIR%\DEMApplication;
CALL :add_app %KRATOS_APP_DIR%\SwimmingDEMApplication;
CALL :add_app %KRATOS_APP_DIR%\CSharpWrapperApplication;
CALL :add_app %KRATOS_APP_DIR%\SolidMechanicsApplication;
CALL :add_app %KRATOS_APP_DIR%\ConstitutiveModelsApplication;
CALL :add_app %KRATOS_APP_DIR%\DelaunayMeshingApplication;
CALL :add_app %KRATOS_APP_DIR%\ContactMechanicsApplication;
CALL :add_app %KRATOS_APP_DIR%\PfemApplication;
CALL :add_app %KRATOS_APP_DIR%\PfemFluidDynamicsApplication;
CALL :add_app %KRATOS_APP_DIR%\PfemSolidMechanicsApplication;
VVVVVVVVVVVVVVVVVVV
REM and it resumes here...
^^^^^^^^^^^^^^^^^^^^^^^^
rem Clean
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\cmake_install.cmake"
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\CMakeCache.txt"
del /F /Q "%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%\CMakeFiles"

rem Configure
@echo on
 cmake                                              ^
 -G"Visual Studio 15 2017 Win64"                    ^
 -H"%KRATOS_SOURCE%"                                ^
 -B"%KRATOS_BUILD%\%KRATOS_BUILD_TYPE%"             ^
 -DBOOST_ROOT="C:\Libraries\boost_1_65_1"           ^
 -DPYTHON_EXECUTABLE="C:\Python36-x64\python.exe"   ^
 -DINCLUDE_FEAST=OFF                                ^
 -DUSE_COTIRE=ON

rem Build
cmake --build "%KRATOS_BUILD%/%KRATOS_BUILD_TYPE%" --target all_unity
goto:eof

:add_app
set KRATOS_APPLICATIONS=%KRATOS_APPLICATIONS%%1;
goto:eof
```

I am missing something? if you check last loogs you can see that only half of that call code is called and then the execution continues to the configuration and compiling part... it makes no sense at all to me :S

pooyan-dadvand commented:
no idea....

Is not entering at all? or not doing what it should do?

roigcarlo committed:
Call sems not to work well in appveyor

roigcarlo committed:
Adapting regular part of cosimulationapp

roigcarlo commented:
@philbucher I updated the cosimulation app, but I think the `ping` & `pong` executables are never copied to the install dir (even in the old cmake) is that intentional? do you want me to move them to some particular place during the install stage?

philbucher commented:
> @philbucher I updated the cosimulation app, but I think the `ping` & `pong` executables are never copied to the install dir (even in the old cmake) is that intentional? do you want me to move them to some particular place during the install stage?

thanks!

Don't worry abt the ping-pong, they are only used for testing. I anyway have some changes related to them, I will fix it alltogether after this PR is merged

roigcarlo committed:
Do not compile runkratos if no debug libs are present

roigcarlo committed:
Merge branch 'master' into cleanup/link_on_install

roigcarlo committed:
Missing D

roigcarlo committed:
Removing duplicate line

roigcarlo committed:
Cleanup configuration files

roigcarlo committed:
Updated scripts and cleanup

roigcarlo commented:
So I tried to document all the options inside the script but was very long and unclear. I added a new entry to the wiki [Compile options](https://github.com/KratosMultiphysics/Kratos/wiki/Compilation-options) with everything and pointed there within the scripts.

If you want I can add the most critical ones inside the scripts themselves. @pooyan-dadvand ?

roigcarlo commented:
Also If I am missing any switch or option please tell me so I'll add it.

roigcarlo committed:
Remove commented code

roigcarlo committed:
Fetching boost root from env

roigcarlo committed:
Why is cuda in the middle of the boost code??

roigcarlo committed:
That does not belong to the core anymore

roigcarlo committed:
Python and boost are now fetched from env variables

roigcarlo committed:
Updated main script for Linux

roigcarlo committed:
Adding install.md

roigcarlo committed:
Grouped some options under the  group

roigcarlo committed:
Removing old code

roigcarlo committed:
Updating readme

roigcarlo committed:
Redundant code

roigcarlo committed:
Moving blas install code to External

roigcarlo commented:
Still trying to find the best way to implement the link on install as a env variable but most of the changes discussed should be there if you want to start reviewing.

roigcarlo committed:
Accidentally removed cotire

roigcarlo committed:
Updating scripts with the newest changes

roigcarlo committed:
More script updates and removed generic linux script as now is the same as the base one

roigcarlo commented:
@KratosMultiphysics/technical-committee I think all the changes we speak about are finally done, so lets start the review

RiccardoRossi commented on CMakeLists.txt:
```diff
@@ -176,28 +196,47 @@ if(NOT DEFINED KRATOS_BUILD_TESTING)
-# Set kratos specific module path
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake_modules")
+# Link on install
+if(NOT DEFINED ENV{KRATOS_INSTALL_PYTHON_USING_LINKS})
```
minor: shouldn't we put all the default handling together at the beginning of the file?

RiccardoRossi commented on CMakeLists.txt:
```diff
@@ -176,28 +196,47 @@ if(NOT DEFINED KRATOS_BUILD_TESTING)
+else(NOT DEFINED ENV{KRATOS_INSTALL_PYTHON_USING_LINKS})
+  SET(INSTALL_PYTHON_USING_LINKS $ENV{KRATOS_INSTALL_PYTHON_USING_LINKS})
+endif(NOT DEFINED ENV{KRATOS_INSTALL_PYTHON_USING_LINKS})
+message("THE VARIABLE VALUE ISSSSSSSS ${INSTALL_PYTHON_USING_LINKS}")
```
this message shuld be either removed or made nicer

roigcarlo replied:
Ouch

RiccardoRossi commented on INSTALL.md:
```diff
@@ -0,0 +1,162 @@
+`PYTHON_EXECUTABLE`
+
+Path to the python executable that Kratos will use. We recommend that you manually set this in case you have multiple versions of python in the system.
+Ubuntu users need to be extra carefull with this as default versions tends to be Python2, while Kratos is preferably compiled with Python3
```
carefuL

RiccardoRossi commented on applications/FreeSurfaceApplication/tests/generalTests.py:
```diff
@@ -19,10 +19,10 @@ def tearDown(self):
         pass
 
     def testSmallExample(self):
-        self.AssertEqual(True, True)
```
ok, even though it does not really belong here

roigcarlo replied:
I just fixed while was testing all the apps in local, if not they didn't pass the tests :S

RiccardoRossi reviewed (commented):
the changes look correct to me, nevertheless if i try to run the default script (adding a few apps) using "sh" instead of "bash" it fails with the message 

 riccardo  cleanup/link_on_install*  ~/Kratos  sh debug_configure.sh 
debug_configure.sh: 12: debug_configure.sh: function: not found
debug_configure.sh: 14: debug_configure.sh: Syntax error: "}" unexpected

roigcarlo committed:
Remove debug message

roigcarlo committed:
Grouping all default setters at the top of the file

roigcarlo committed:
Typo

roigcarlo commented:
Thx @RiccardoRossi, made all the changes!

RiccardoRossi commented:
ok, so i am giving my approval, as agreed wait to have more than one

RiccardoRossi reviewed (dismissed):
good for me.

philbucher commented:
I am looking at it right now, can you wait with merging until after our meeting this afternoon?

philbucher commented:
> the changes look correct to me, nevertheless if i try to run the default script (adding a few apps) using "sh" instead of "bash" it fails with the message
> 
> riccardo cleanup/link_on_install* ~/Kratos sh debug_configure.sh
> debug_configure.sh: 12: debug_configure.sh: function: not found
> debug_configure.sh: 14: debug_configure.sh: Syntax error: "}" unexpected

same here, how did you fix it?

roigcarlo committed:
Reducing the list of default apps

roigcarlo commented:
Seems to be because old versions of sh do not like functions. I am looking for an alternative, if it keeps failing I will remove that.

Should work with "bash" instead of "sh"

philbucher commented:
> Seems to be because old versions of sh do not like functions. I am looking for an alternative, if it keeps failing I will remove that.
> 
> Should work with "bash" instead of "sh"

ok thx will try

FYI I use the following for "functions" in my `sh` scripts:
~~~
ompsetthreads() {
    if [[ $# = 0 ]]; then
        export OMP_NUM_THREADS=1 && echo "No input, setting 1 thread"
    else
        export OMP_NUM_THREADS=$1 && echo "Set $1 OpenMP thread"
    fi
}
export -f ompsetthreads
~~~

philbucher commented:
some initial comments:
- the default build-dir is not added to the gitignore, do you think it makes sense adding it?
- why did you change the install dir? Why no longer `KratosMultiphysics` on the top-level? (see pic)
- why are the tests in different directories after the install? (see pic)
- the `INSTALL.md` is not pointing to the `README.md`s of the apps (I think we discussed to do this)

![image](https://user-images.githubusercontent.com/25484702/69532338-dc4f5b80-0f75-11ea-83b0-f9c82b02b241.png)
This seems inconvenient and confusing to me, is there a technical reason?

roigcarlo committed:
Correcting error in the ignore file

philbucher committed:
removed print

philbucher committed:
Merge remote-tracking branch 'origin/cleanup/link_on_install' into cleanup/link_on_install

philbucher commented:
more observations:
- I get the warning `Unknown CMAKE_BUILD_TYPE '', using 'Release'.` even though I was using the default configure-script, where the build-type is defined
- Do we really need to issue the message abt ALL the installed files? This is a lot, can easily flood the terminal
- How do we handle custom-configure scripts? Shoudn't we put them in a folder? I fear that ppl will change this file, commit to it and then push it (same happens with the `.gitignore`)

philbucher committed:
correcting gitignore => STILL HAS TO BE CLEANED!

roigcarlo committed:
Fixed bug after changig defaults order

roigcarlo commented:
Hi @philbucher 

> the default build-dir is not added to the gitignore, do you think it makes sense adding it?

Typo in the .gitignore, was there but in a initial implementation the folder was hidden and I had my local ignore already ignoring that. Fixed now.

> why did you change the install dir? Why no longer KratosMultiphysics on the top-level? (see pic)

Not sure to understand this,
The top install level remains the same. The Linux / Release folders are just there in order to organize the install dir and avoid clashes.

> why are the tests in different directories after the install? (see pic)

Test are in the application folder as always. Are they in a different location?

> the INSTALL.md is not pointing to the README.mds of the apps (I think we discussed to do this)

I tough we agreed on removing it but didn't get the pointing to the other Readme. Doing it now

> I get the warning Unknown CMAKE_BUILD_TYPE '', using 'Release'. even though I was using the default configure-script, where the build-type is defined

My bad, left a line while fixing some of @RiccardoRossi suggestions

> How do we handle custom-configure scripts?

They should copy the script if they want to make changes, to be honest is more ore less what you are doing now anyway but instead of just ignoring the folder they have to ignore a script. For now I would leave it like that, people on Altair doesn't seem to have problems with it (@pooyan-dadvand is that right?). If we see that gives problems or people is constantly pushing it we can safely move that to a folder as we use to have.

roigcarlo committed:
Merge branch 'cleanup/link_on_install' of https://github.com/KratosMultiphysics/Kratos into cleanup/link_on_install

roigcarlo commented:
I forgot the prints. Yes they can be removed if we want, there are flags for that, we can even decide if we want to keep the install ones or only the up-to-date etc...

If you all agree its easy to do

roigcarlo committed:
.gitignore cleanup

philbucher commented:
ok thx for the answers.
Yes the tests are in different dirs:
![image](https://user-images.githubusercontent.com/25484702/69534899-fccde480-0f7a-11ea-89ed-6ae700489ccc.png)


I understand that you have the `bin` for the different builds, but I would say we don't need the OS-named folder. Or do you have a usecase for this?

philbucher commented:
Also I cannot figure out the following error:
~~~
  File "test_StructuralMechanicsApplication.py", line 2, in <module>
    import KratosMultiphysics
  File "/home/philippb/software/Kratos_prod/bin/Linux/Release/KratosMultiphysics/__init__.py", line 16, in <module>
    from Kratos import *
ImportError: libKratosCore.so: cannot open shared object file: No such file or directory
~~~

I tried adding the `libs` folder to the `PATH`, to the `PYTHONPATH` and also compiled with `CMAKE_INSTALL_RPATH_USE_LINK_PATH=TRUE`, always same error

do you have a clue?

Before it was sufficient for me to use `CMAKE_INSTALL_RPATH_USE_LINK_PATH=TRUE`

roigcarlo committed:
Pointing to applications with install instructions

roigcarlo commented:
> I understand that you have the bin for the different builds, but I would say we don't need the OS-named folder. Or do you have a usecase for this?

I do, but probably is to niche, I will change it :+1: 

> Yes the tests are in different dirs:

Still can't see the problem with the tests. There are exactly were they are supposed to be :S I didn't change any of that. Do you want them to be all in the same folder?

> do you have a clue?

You should point `LD_LIBRARY_PATH=/home/philippb/software/Kratos_prod/bin/Linux/Release/libs`
`PATH` is only for win, if you do it on linux only is used to search for binaries.

`CMAKE_INSTALL_RPATH_USE_LINK_PATH` should only affect runkratos or the dependencies of the library itself, in this case has no effect.

roigcarlo committed:
Remove system name from default install dir

roigcarlo committed:
And correcting doc accordingly

philbucher commented:
Ok tests locations are fine I guess, however I suggest the following folder structure:
~~~
- Release
    - KratosMultiphysics
    - libs
    - python_tests
        - applications
        - kratos
~~~

Even with the LD_LIB... it cannot find the C-Libs

Note that it is working with the old mechanism, I only use `CMAKE_INSTALL_RPATH_USE_LINK_PATH` and do not modify the LD_LIB_PATH...

roigcarlo commented:
I think if you do not modify the LD_LIBRARY_PATH and use CMAKE_INSTALL_RPATH_USE_LINK_PATH the RPATH of the object will be malformed. 

Can you `readelf -d /home/philippb/software/Kratos_prod/bin/Linux/Release/libs/libKratosCore.so`?

pooyan-dadvand commented:
> ok thx for the answers.
> Yes the tests are in different dirs:
> I understand that you have the `bin` for the different builds, but I would say we don't need the OS-named folder. Or do you have a usecase for this?

It simplifies the deployment of nightly builds and the weekly regression tests. But I understand that it may complicate the local use.

philbucher commented:
> I think if you do not modify the LD_LIBRARY_PATH and use CMAKE_INSTALL_RPATH_USE_LINK_PATH the RPATH of the object will be malformed.
> 
> Can you `readelf -d /home/philippb/software/Kratos_prod/bin/Linux/Release/libs/libKratosCore.so`?

~~~
philippb:~/software/Kratos_prod$ readelf -d /home/philippb/software/Kratos_prod/bin/Linux/Release/libs/libKratosCore.so

Dynamic section at offset 0x1ea17d8 contains 31 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgomp.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000e (SONAME)             Library soname: [libKratosCore.so]
 0x000000000000000c (INIT)               0x45dde8
 0x000000000000000d (FINI)               0x16ae2a8
 0x0000000000000019 (INIT_ARRAY)         0x205d9f8
 0x000000000000001b (INIT_ARRAYSZ)       1784 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x205e0f0
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x1f0
 0x0000000000000005 (STRTAB)             0xc2e48
 0x0000000000000006 (SYMTAB)             0x304d8
 0x000000000000000a (STRSZ)              2701881 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x20ab000
 0x0000000000000002 (PLTRELSZ)           111864 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x4428f0
 0x0000000000000007 (RELA)               0x362e08
 0x0000000000000008 (RELASZ)             916200 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0x362bf8
 0x000000006fffffff (VERNEEDNUM)         6
 0x000000006ffffff0 (VERSYM)             0x356882
 0x000000006ffffff9 (RELACOUNT)          539
 0x0000000000000000 (NULL)               0x0
~~~

roigcarlo commented:
Looks normal.

`objdump -x libs/libKratosCore.so | grep RPATH`?

philbucher commented:
> Looks normal.
> 
> `objdump -x libs/libKratosCore.so | grep RPATH`?

nothing :/

this is still with `DCMAKE_INSTALL_RPATH_USE_LINK_PATH`, should I compile without?

roigcarlo commented:
Yup, much better. I dropped that some years ago. To be honest I didn't know anyone was using it :S

roigcarlo committed:
Fixing build type

roigcarlo committed:
Merge branch 'master' into cleanup/link_on_install

roigcarlo committed:
New path without SO

roigcarlo committed:
Fixing problem with advanced shell functions

roigcarlo commented:
@philbucher @RiccardoRossi Does the changes in the script work with sh?

roigcarlo committed:
renaming to standard_

philbucher commented:
Now it compiles with `sh`
However ti still issues the installed files

philbucher commented:
I still couldn't figure out how to make it run, same error as before, but this should not be the blocker here

roigcarlo committed:
Adding defaults to some variable to get from env

roigcarlo committed:
Typo and moving to scripts

roigcarlo commented:
Ok I added the defaults as @pooyan-dadvand wanted (please double check) and moved the scrips to the scripts folder so now they are located in `scripts/standard_configure.xxx`

Also the `KRATOS_SOURCE` variable is initialized so it takes the location of the script by default (it does not matter from where you call the script, it will point to the folder below where it is located)

Please let the win one run because I am not sure we can use that.

philbucher commented:
@roigcarlo I think you made a mistake, seems like moving the configure-scripts to `scripts` did not work

roigcarlo commented:
Hum true, the win one is still there...

roigcarlo committed:
Moving scripts again

roigcarlo commented:
Ok now

philbucher commented:
ok now seems to be back

so what should be do by default? Copy the config-script to a folder or what do you suggest?

roigcarlo commented:
well now you can copy the script wherever you want and will work. If you copy paste it inside the folder you only have to add/remove the apps (and python) and the rest is automatic.

If I can I will try to find how to autodetect python3 but no promises

philbucher reviewed (dismissed):
ok to me, the details can be done in future PRs if necessary

nice work :+1:

roigcarlo committed:
Win changes: Default python, added blas / lapack, default source

roigcarlo committed:
Typo

roigcarlo committed:
Missing dependency

roigcarlo committed:
Merge branch 'cleanup/link_on_install' of https://github.com/KratosMultiphysics/Kratos into cleanup/link_on_install

roigcarlo committed:
Fixing if, removing duplicated message

roigcarlo committed:
Typo in windows build dir

roigcarlo commented:
Ok. Should we merge? Someone can reaprove?

RiccardoRossi reviewed (approved):
well done!

philbucher commented:
nice job!

do you still need the branch?

roigcarlo commented:
Let's keep it til Monday
